cpp/include/cudf/join.hpp

```diff
@@ -213,6 +213,85 @@ std::unique_ptr<cudf::experimental::table> full_join(
                          std::vector<std::pair<cudf::size_type, cudf::size_type>> const& columns_in_common,
                          rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource());
 
+/**
+ * @brief  Performs an inner join using the nested loop join implementation.
+ *
+ * Nested loop join computes the cross product of the two tables (compares
+ * all rows in Left to all rows in Right) and filters the product based upon
+ * a provided filtering function which identifies the rows that will be part
+ * of the output join.
+ *
+ * @example Left a: {0, 1, 2}
+ *          Right b: {1, 2, 3}, a: {1, 2, 5}
+ *          Filter compares Left[0] to Right[1], returns true if they match
+ *          columns_in_common: { {0, 1} }
+ * Result: { a: {1, 2}, b: {1, 2} }
+ *
+ * @example Left a: {0, 1, 2}
+ *          Right b: {1, 2, 3}, c: {1, 2, 5}
+ *          Filter compares Left[0] to Rignt[0], returns true if they match
+ *          columns_in_common: { }
+ * Result: { a: {1, 2}, b: {1, 2}, c: {1, 2} }
+ *
+ * @throws cudf::logic_error if `columns_in_common` contains a pair of indices
+ * (L, R) if L does not exist in `left_on` or R does not exist in `right_on`.
+ * @throws cudf::logic_error if `columns_in_common` contains a pair of indices
+ * (L, R) such that the type of `L` within `left_on` is not equal to
+ * type of R within `right_on`
+ * @throws cudf::logic_error if number of elements in `left_on` or `right_on`
+ * mismatch.
+ * @throws std::out_of_range if element of `left_on` or `right_on` exceed the
+ * number of columns in the left or right table.
+ *
+ * @param[in] left               The left table
+ * @param[in] right              The right table
+ * @param[in] columns_in_common  A vector of pairs of column indices into
+ * `left` and `right`, respectively, that are "in common". For "common"
+ * columns, only a single output column will be produced, which is gathered
+ * from the `left` table.
+ * @param[in] filter             A function that accepts a left index and
+ *                               right index corresponding to rows in the
+ *                               left and right tables.  Returns true if
+ *                               the row is part of the resulting join, false
+ *                               if not
+ * @param mr                     Memory resource used to allocate the returned
+ *                               table and columns
+ *
+ * @returns    Result of joining `left` and `right` tables, filtered by the
+ *             provided filter. The resulting table will be joined columns of
+ *             `left(including common columns)+right(excluding common columns)`.
+ */
+template <typename Filter>
+std::unique_ptr<table> nested_loop_inner_join(cudf::table_view const& left,
+                                              cudf::table_view const& right,
+                                              std::vector<std::pair<cudf::size_type, cudf::size_type>> const& columns_in_common,
```

Comment:
What do you think about having overloads that take a vector of string pairs here (most likely outside of the scope of this PR), so user can specify the column names? To me it seems more convenient than specifying the indices.

Reply:
I wouldn't be opposed.  However, I think that's an overall library API discussion.  AFAIK all of the APIs in cuDF use the column indices.

Reply:
Columns don't have names. libcudf functions expressly do not operate on or use names.